### ISSUE TITLE
Allow for custom normalization of schemas in exports of generators

### ISSUE THREAD

**Issue by BartKoppelmans** [nrwl/nx#12364]
TL;DR: Using the exported generator in a local plugin or generator means that developers are dependent on and limited by the schema normalization as specified by NX. 

## Description & Motivation
The publicly exported generators are implemented such that the raw input always passes through the "default" normalization. When implementing a local plugin, typically (read: in my case) it is desired for developers to have more control over the input that enters the actual generator, without being limited by the schema normalization. 

While schema normalization is required and highly beneficial in the CLI, this also means that local plugins are limited to this CLI behaviour, which is generally aimed at providing as little as possible into a generator and figuring out the rest. 

An example: when using the library generator in composition (e.g. in a local plugin), the developer cannot specify both the destination and project name separately. In my case, the workspace is structured differently than the suggested structure by NX. This means that the `workspaceLayout` in nx.json is set to put everything in the root. By using the exported generators, this means that all project names automatically start with `libs` if I would still want to put them in such a directory, since it goes through the normalization.

While this is definitely desired behaviour when using generators through the CLI, it limits developers willing to go the extra mile by implementing a local plugin, which desires a different behaviour. Currently the only way I found to tackle this problem was to create a copy of the implementation using the not-publicly exported generator functions or by creating a local implementation, both completely counteracting the desire to "compose" generators.

## Suggested Implementation
Most likely the easiest implementation would be to split the implementation of generators into stages: (i) validation, (ii) normalization, and (iii) implementation. Each of these could be exported publicly, with the default export being the combination of the three stages. Of course these stages are simply a suggestion, where e.g. validation and normalization might still be combined.

An example of the move generator following this structure:

```ts
export async function moveGenerator(tree: Tree, raw: Schema) {
  const projectConfig = readProjectConfiguration(tree, raw.projectName)
  
  await moveGeneratorValidationStep(tree, raw, projectConfig)
  const schema = await moveGeneratorNormalizationStep(tree, raw, projectConfig)
  await moveGeneratorImplementationStep(tree, schema, projectConfig)
}

export async function moveGeneratorValidationStep(
  tree: Tree,
  raw: Schema,
  projectConfig = readProjectConfiguration(tree, raw.projectName)
): Promise<void> {
  checkDestination(tree, raw, projectConfig)
}

export async function moveGeneratorNormalizationStep(
  tree: Tree,
  raw: Schema,
  projectConfig = readProjectConfiguration(tree, raw.projectName)
): Promise<NormalizedSchema> {
  return normalizeSchema(tree, raw, projectConfig)
}

export async function moveGeneratorImplementationStep(
  tree: Tree,
  schema: NormalizedSchema,
  projectConfig = readProjectConfiguration(tree, schema.projectName)
): Promise<void> {
  moveProjectConfiguration(tree, schema, projectConfig)
  moveProject(tree, schema, projectConfig)
  updateImports(tree, schema, projectConfig)
  updateProjectRootFiles(tree, schema, projectConfig)
  updateCypressConfig(tree, schema, projectConfig)
  updateJestConfig(tree, schema, projectConfig)
  updateStorybookConfig(tree, schema, projectConfig)
  updateEslintrcJson(tree, schema, projectConfig)
  updateReadme(tree, schema)
  updatePackageJson(tree, schema)
  updateBuildTargets(tree, schema)
  updateDefaultProject(tree, schema)
  updateImplicitDependencies(tree, schema)

  // Depending on desired behaviour, this might also be moved to the default export.
  if (!schema.skipFormat) {
    await formatFiles(tree)
  }
}

export default moveGenerator
```

## Alternate Implementations
Another way that similar behaviour could be achieved, is by adding the normalized properties to the public schema, and then in the normalization steps using either the specified property, or the own implementation of the normalization for that property. However, this is bound to result in issues with properties having the same name.

## Quick notes
1. The title definitely does not cover the feature request/issue well. Feel free to change it as you wish. 
2. I can understand that while this might seem to be an easy change in implementation, there are probably generators that follow a different structure or cannot be split. If needed, I'd enjoy helping in refactoring the generators, depending on your desired implementation. AKA:
- [x] I'd be willing to help implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))
